packages/@aws-cdk/core/lib/asset-staging.ts

```diff
@@ -204,37 +262,60 @@ export class AssetStaging extends Construct {
     return bundleDir;
   }
 
-  private calculateHash(props: AssetStagingProps): string {
-    let hashType: AssetHashType;
-
-    if (props.assetHash) {
-      if (props.assetHashType && props.assetHashType !== AssetHashType.CUSTOM) {
-        throw new Error(`Cannot specify \`${props.assetHashType}\` for \`assetHashType\` when \`assetHash\` is specified. Use \`CUSTOM\` or leave \`undefined\`.`);
+  /**
+   * Determines the hash type from user-given prop values.
+   *
+   * @param assetHashType Asset hash type construct prop
+   * @param assetHash Asset hash given in the construct props
+   */
+  private determineHashType(assetHashType?: AssetHashType, assetHash?: string) {
+    if (assetHash) {
+      if (assetHashType && assetHashType !== AssetHashType.CUSTOM) {
+        throw new Error(`Cannot specify \`${assetHashType}\` for \`assetHashType\` when \`assetHash\` is specified. Use \`CUSTOM\` or leave \`undefined\`.`);
       }
-      hashType = AssetHashType.CUSTOM;
-    } else if (props.assetHashType) {
-      hashType = props.assetHashType;
+      return AssetHashType.CUSTOM;
+    } else if (assetHashType) {
+      return assetHashType;
     } else {
-      hashType = AssetHashType.SOURCE;
+      return AssetHashType.SOURCE;
     }
+  }
 
+  private calculateHash(hashType: AssetHashType, assetHash?: string, bundling?: BundlingOptions): string {
```

Comment:
Seems like the SOURCE and CUSTOM cases share most of the logic. Perhaps we can write it like this to simplify:

```ts
if (hashType === BUNDLE) {
  return fingerprint(bundleDir, ...);
}

const hash = new Array<any>(); // hash components

// if asset hash is provided by user, use it, otherwise fingerprint the source
hash.push(assetHash ?? fingerprint(sourcePath, ...));

// include bundling configuration in hash
if (bundling) {
  hash.push(JSON.stringify(bundling));
}

return crypto.createHash('sha256').update(JSON.stringify(hash)).digest('hex');
```

Reply:
I've simplified the logic here by quite a bit with your feedback, but I had to deviate a bit from this suggestion. There seemed to be a lot of test code that checked for specific result asset hashes. I was concerned about making sweeping changes across the project, so I worked-in some compatibility. I hope this is acceptable.